File: tools/validate-features.js
/**
 * validate-features.js
 * Validates D&D 5e SRD feature/effect data integrity.
 * Usage: node tools/validate-features.js
 */

const fs = require('fs');
const path = require('path');

const DATA_DIR = path.join(__dirname, '..', 'data');
const EFFECTS_DIR = path.join(DATA_DIR, 'effects');
const DESCRIPTIONS_DIR = path.join(DATA_DIR, 'descriptions');
const CLASSES_DIR = path.join(DATA_DIR, 'classes');

const KNOWN_TYPES = new Set([
  'lookup', 'none',
  'proficiency', 'choice', 'innate', 'mainspell', 'cantrips',
  'resistance', 'immunity', 'vulnerability',
  'savingThrow', 'stat', 'speed', 'vision', 'language', 'feat', 'maxHP'
]);

const HANDLED_IN_GLOSSARY = new Set([
  'innate', 'proficiency', 'vision', 'speed', 'choice',
  'cantrips', 'resistance', 'immunity', 'vulnerability',
  'savingThrow', 'stat', 'language', 'feat', 'mainspell'
]);

const CHOICE_TYPE_MAP = {
  invocation: 'class-effects.json:eldritch invocations',
  discipline: 'subclass-effects.json:disciple-of-the-elements',
  maneuver: 'subclass-effects.json:combat-superiority'
};

let errors = [];
let warnings = [];

function readJSON(relPath) {
  const p = path.join(DATA_DIR, relPath);
  if (!fs.existsSync(p)) return null;
  return JSON.parse(fs.readFileSync(p, 'utf-8'));
}

function readEffectsFile(filename) {
  const p = path.join(EFFECTS_DIR, filename);
  if (!fs.existsSync(p)) return {};
  return JSON.parse(fs.readFileSync(p, 'utf-8'));
}

function readDescriptionFile(filename) {
  const p = path.join(DESCRIPTIONS_DIR, filename);
  if (!fs.existsSync(p)) return {};
  return JSON.parse(fs.readFileSync(p, 'utf-8'));
}

function readClassFile(filename) {
  const p = path.join(CLASSES_DIR, filename);
  if (!fs.existsSync(p)) return null;
  return JSON.parse(fs.readFileSync(p, 'utf-8'));
}

// 1. Load all effect data
const classEffects = readEffectsFile('class-effects.json');
const subclassEffects = readEffectsFile('subclass-effects.json');
const raceEffects = readEffectsFile('race-effects.json');
const featEffects = readEffectsFile('feat-effects.json');
const classOptionEffects = readEffectsFile('class-option-effects.json');

const classEffectsEntries = classEffects.effects || {};
const subclassEffectsEntries = subclassEffects.effects || {};
const raceEffectsEntries = raceEffects.effects || {};
const featEffectsEntries = featEffects.effects || {};

// 2. Check all effect types are known
console.log('\n=== CHECK 1: Unknown effect types ===');
[classEffectsEntries, subclassEffectsEntries, raceEffectsEntries, featEffectsEntries].forEach((entries, i) => {
  const source = ['class-effects.json', 'subclass-effects.json', 'race-effects.json', 'feat-effects.json'][i];
  Object.entries(entries).forEach(([key, val]) => {
    if (!KNOWN_TYPES.has(val.type)) {
      errors.push(`${source}:${key} has unknown type "${val.type}"`);
    }
  });
});
console.log(`  Checked types in 4 effect files`);

// 3. Check glossary coverage for non-lookup types
console.log('\n=== CHECK 2: Glossary handler coverage ===');
[classEffectsEntries, subclassEffectsEntries, raceEffectsEntries].forEach((entries, i) => {
  const source = ['class-effects.json', 'subclass-effects.json', 'race-effects.json'][i];
  const nonLookup = Object.entries(entries).filter(([_, v]) => v.type !== 'lookup' && v.type !== 'none');
  const uncovered = nonLookup.filter(([_, v]) => !HANDLED_IN_GLOSSARY.has(v.type));
  uncovered.forEach(([key, val]) => {
    warnings.push(`${source}:${key} has type "${val.type}" which has no handler in GlossaryUI.renderEffectDetails`);
  });
});
console.log(`  ${warnings.filter(w => w.includes('no handler')).length} uncovered types found`);

// 4. Check choice options cross-reference
console.log('\n=== CHECK 3: Choice options cross-reference ===');
const allEffectKeys = new Set([
  ...Object.keys(classEffectsEntries),
  ...Object.keys(subclassEffectsEntries),
  ...Object.keys(raceEffectsEntries),
  ...Object.keys(featEffectsEntries)
]);

[classEffectsEntries, subclassEffectsEntries].forEach((entries, i) => {
  const source = ['class-effects.json', 'subclass-effects.json'][i];
  Object.entries(entries).forEach(([key, val]) => {
    if (val.type === 'choice' && val.options) {
      val.options.forEach(opt => {
        const optKey = opt.toLowerCase().replace(/'/g, '').replace(/[^a-z0-9]+/g, '-').replace(/^-|-$/g, '');
        const hasEffect = allEffectKeys.has(optKey) || key === optKey;
        if (!hasEffect && val.choiceType !== 'hunter-prey' && val.choiceType !== 'hunter-defense' && val.choiceType !== 'hunter-attack' && val.choiceType !== 'hunter-superior' && val.choiceType !== 'totem-spirit' && val.choiceType !== 'totem-attunement' && val.choiceType !== 'resistance-type' && val.choiceType !== 'fighting-style' && val.choiceType !== 'transmuter-stone' && val.choiceType !== 'transmuter-master' && val.choiceType !== 'third-eye') {
          warnings.push(`${source}:${key} option "${opt}" (key: ${optKey}) has no matching effect entry`);
        }
      });
    }
  });
});

// 5. Check level prereq completeness
console.log('\n=== CHECK 4: Level prerequisite completeness ===');
[classEffectsEntries, subclassEffectsEntries].forEach((entries, i) => {
  const source = ['class-effects.json', 'subclass-effects.json'][i];
  Object.entries(entries).forEach(([key, val]) => {
    if (val.type === 'choice' && val.levelPrereqs && val.options) {
      val.options.forEach(opt => {
        if (val.levelPrereqs[opt] === undefined) {
          warnings.push(`${source}:${key} option "${opt}" has no level prerequisite defined`);
        }
      });
    }
  });
});

// 6. Check for duplicate keys across files
console.log('\n=== CHECK 5: Duplicate effect keys ===');
const keySources = {};
[classEffectsEntries, subclassEffectsEntries, raceEffectsEntries].forEach((entries, i) => {
  const source = ['class-effects.json', 'subclass-effects.json', 'race-effects.json'][i];
  Object.keys(entries).forEach(key => {
    if (!keySources[key]) keySources[key] = [];
    keySources[key].push(source);
  });
});
Object.entries(keySources).forEach(([key, sources]) => {
  if (sources.length > 1) {
    warnings.push(`Effect key "${key}" appears in multiple files: ${sources.join(', ')}`);
  }
});

// 7. Check feature names in class JSONs map to effect keys
console.log('\n=== CHECK 6: Class JSON feature to effect mapping ===');
const classFiles = fs.readdirSync(CLASSES_DIR).filter(f => f.endsWith('.json'));
classFiles.forEach(cf => {
  const cls = readClassFile(cf);
  if (!cls) return;
  const clsId = cls.id || path.basename(cf, '.json');
  const featuresByLevel = cls.features || {};
  Object.entries(featuresByLevel).forEach(([lvl, data]) => {
    (data.features || []).forEach(fname => {
      const fkey = fname.toLowerCase();
      if (!allEffectKeys.has(fkey) && fkey !== 'spellcasting' && fkey !== 'pact magic' && fkey !== '+1 feat') {
        warnings.push(`${cf}: Level ${lvl} feature "${fname}" (key: ${fkey}) has no matching effect entry`);
      }
    });
  });
});

// 8. Report
console.log('\n========================================');
console.log(`VALIDATION COMPLETE`);
console.log(`  Errors:   ${errors.length}`);
console.log(`  Warnings: ${warnings.length}`);
console.log('========================================\n');

if (errors.length > 0) {
  console.log('=== ERRORS ===');
  errors.forEach(e => console.log(`  [ERROR] ${e}`));
}

if (warnings.length > 0) {
  console.log('=== WARNINGS ===');
  warnings.forEach(w => console.log(`  [WARN] ${w}`));
}

if (errors.length === 0 && warnings.length === 0) {
  console.log('All checks passed!');
}

File: tools/validate-features.py
#!/usr/bin/env python3
"""
validate-features.py — validates D&D 5e SRD feature/effect data integrity.
Usage: python3 tools/validate-features.py
"""

import json
import os
import re

DATA_DIR = os.path.join(os.path.dirname(__file__), '..', 'data')
EFFECTS_DIR = os.path.join(DATA_DIR, 'effects')
DESCRIPTIONS_DIR = os.path.join(DATA_DIR, 'descriptions')
CLASSES_DIR = os.path.join(DATA_DIR, 'classes')

KNOWN_TYPES = {
    'lookup', 'none',
    'proficiency', 'choice', 'innate', 'mainspell', 'cantrips',
    'resistance', 'immunity', 'vulnerability',
    'savingThrow', 'stat', 'speed', 'vision', 'language', 'feat', 'maxHP', 'skill'
}

HANDLED_IN_GLOSSARY = {
    'innate', 'proficiency', 'vision', 'speed', 'choice',
    'cantrips', 'resistance', 'immunity', 'vulnerability',
    'savingThrow', 'stat', 'language', 'feat', 'mainspell',
    'maxHP', 'skill'
}

OPTIONAL_CHOICE_TYPES = {
    'hunter-prey', 'hunter-defense', 'hunter-attack', 'hunter-superior',
    'totem-spirit', 'totem-attunement', 'resistance-type',
    'fighting-style', 'transmuter-stone', 'transmuter-master', 'third-eye'
}

errors = []
warnings = []

def read_json(path):
    if not os.path.exists(path):
        return None
    with open(path) as f:
        return json.load(f)

def option_to_key(name):
    return re.sub(r'[^a-z0-9]+', '-', name.lower().replace("'", '')).strip('-')

# 1. Load all effect data
class_effects = (read_json(os.path.join(EFFECTS_DIR, 'class-effects.json')) or {}).get('effects', {})
subclass_effects = (read_json(os.path.join(EFFECTS_DIR, 'subclass-effects.json')) or {}).get('effects', {})
race_effects = (read_json(os.path.join(EFFECTS_DIR, 'race-effects.json')) or {}).get('effects', {})
feat_effects = (read_json(os.path.join(EFFECTS_DIR, 'feat-effects.json')) or {}).get('effects', {})
class_option_effects = (read_json(os.path.join(EFFECTS_DIR, 'class-option-effects.json')) or {}).get('options', {})

all_effect_keys = set()
all_effect_keys.update(class_effects.keys())
all_effect_keys.update(subclass_effects.keys())
all_effect_keys.update(race_effects.keys())
all_effect_keys.update(feat_effects.keys())

effect_files = [
    ('class-effects.json', class_effects),
    ('subclass-effects.json', subclass_effects),
    ('race-effects.json', race_effects),
    ('feat-effects.json', feat_effects),
]

# CHECK 1: Unknown types
print('\n=== CHECK 1: Unknown effect types ===')
for source, entries in effect_files:
    for key, val in entries.items():
        t = val.get('type')
        if t not in KNOWN_TYPES:
            errors.append(f'{source}:{key} has unknown type "{t}"')
print(f'  Checked types in {len(effect_files)} effect files')

# CHECK 2: Glossary handler coverage
print('\n=== CHECK 2: Glossary handler coverage ===')
for source, entries in effect_files:
    for key, val in entries.items():
        t = val.get('type')
        if t not in ('lookup', 'none') and t not in HANDLED_IN_GLOSSARY:
            warnings.append(f'{source}:{key} has type "{t}" with no glossary handler')
print(f'  {len([w for w in warnings if "glossary handler" in w])} uncovered type(s)')

# CHECK 3: Choice options cross-reference
print('\n=== CHECK 3: Choice options cross-reference ===')
for source, entries in effect_files[:2]:
    for key, val in entries.items():
        if val.get('type') == 'choice' and val.get('options'):
            ct = val.get('choiceType', '')
            for opt in val['options']:
                opt_key = option_to_key(opt)
                if opt_key not in all_effect_keys and ct not in OPTIONAL_CHOICE_TYPES:
                    warnings.append(f'{source}:{key} option "{opt}" (key: {opt_key}) has no matching effect entry')
print(f'  Cross-reference check done')

# CHECK 4: Level prereq completeness
print('\n=== CHECK 4: Level prerequisite completeness ===')
for source, entries in effect_files[:2]:
    for key, val in entries.items():
        if val.get('type') == 'choice' and val.get('levelPrereqs') and val.get('options'):
            for opt in val['options']:
                if opt not in val['levelPrereqs']:
                    warnings.append(f'{source}:{key} option "{opt}" has no level prerequisite defined')
print(f'  Level prereq check done')

# CHECK 5: Duplicate keys
print('\n=== CHECK 5: Duplicate effect keys ===')
key_sources = {}
for source, entries in effect_files[:3]:
    for key in entries:
        key_sources.setdefault(key, []).append(source)
for key, sources in key_sources.items():
    if len(sources) > 1:
        warnings.append(f'Effect key "{key}" in multiple files: {", ".join(sources)}')
print(f'  Duplicate key check done')

# CHECK 6: Class JSON feature to effect mapping
print('\n=== CHECK 6: Class JSON feature to effect mapping ===')
if os.path.isdir(CLASSES_DIR):
    for cf in sorted(os.listdir(CLASSES_DIR)):
        if not cf.endswith('.json'):
            continue
        cls = read_json(os.path.join(CLASSES_DIR, cf))
        if not cls:
            continue
        features_by_level = cls.get('features', {})
        for lvl, data in features_by_level.items():
            for fname in data.get('features', []):
                fkey = fname.lower()
                if fkey not in all_effect_keys and fkey not in ('spellcasting', 'pact magic', '+1 feat', '+1 proficiency'):
                    warnings.append(f'{cf}: Level {lvl} feature "{fname}" (key: {fkey}) has no matching effect entry')
print(f'  Class feature mapping check done')

# CHECK 7: Option descriptions for choice options
print('\n=== CHECK 7: Description coverage for choice options ===')
maneuvers_desc = read_json(os.path.join(DESCRIPTIONS_DIR, 'maneuvers.json')) or {}
invocations_desc = read_json(os.path.join(DESCRIPTIONS_DIR, 'invocations.json')) or {}
disciplines_desc = read_json(os.path.join(DESCRIPTIONS_DIR, 'disciplines.json')) or {}
# Check invocation options have descriptions
inv_effect = class_effects.get('eldritch invocations', {})
if inv_effect.get('options'):
    for opt in inv_effect['options']:
        opt_key = option_to_key(opt)
        if opt not in invocations_desc:
            warnings.append(f'Invocation "{opt}" has no description in invocations.json')
# Check discipline options have descriptions
disc_effect = subclass_effects.get('disciple-of-the-elements', {})
if disc_effect.get('options'):
    for opt in disc_effect['options']:
        if opt not in disciplines_desc:
            warnings.append(f'Discipline "{opt}" has no description in disciplines.json')

print(f'  Description coverage check done')

# REPORT
print('\n' + '=' * 40)
print(f'VALIDATION COMPLETE')
print(f'  Errors:   {len(errors)}')
print(f'  Warnings: {len(warnings)}')
print('=' * 40)

if errors:
    print('\n=== ERRORS ===')
    for e in errors:
        print(f'  [ERROR] {e}')

if warnings:
    print('\n=== WARNINGS ===')
    for w in warnings:
        print(f'  [WARN] {w}')

if not errors and not warnings:
    print('  All checks passed!')
